### PR TITLE
feat: add abby cli

### PR DESCRIPTION
**Pull request by cstrnt**
This PR includes the following:

- Adds a CLI to use interact with the API to create tests, environments and flags
- Adds API keys to A/BBY
- Adds a section on the /profile page to create and revoke API keys
- Adds new endpoints to interact with A/BBY
- Adds a `defineConfig` helper function to all integrations + the core package
- Adds docs for the CLI

### CODE REVIEW

File: apps/web/src/components/CreateAPIKeyModal.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        try {
+          setName("");
+          toast.success("API Key created");
+          onClose();
+        } catch (e) {
+          toast.error(
+            e instanceof TRPCClientError &&
+              e.shape.code === TRPC_ERROR_CODES_BY_KEY.FORBIDDEN
+              ? e.message
+              : "Error creating api Key"
```
Comment (slaesh) [inline apps/web/src/components/CreateAPIKeyModal.tsx:47]:
```suggestion
              : "Error creating API Key"
```

File: apps/web/prisma/schema.prisma [hunk 2]
```diff
@@ -256,3 +256,17 @@ model CouponCodes {
 
   @@index([redeemedById])
 }
+
+model ApiKey {
+  id        String   @id @default(cuid())
+  name      String
+  hashedKey String   @unique
+  validDays Int
```
Comment (slaesh) [inline apps/web/prisma/schema.prisma:264]: why in days? why not just an "end-date" validUntil or similar, so you wont calculate over and over again? even the database could easily filter for us..

File: apps/angular-example/src/app/abby.ts [hunk 3]
```diff
@@ -5,6 +5,7 @@ import { environment } from "../environments/environment";
 export const abby = {
   projectId: environment.ABBY_PROJECT_ID,
   currentEnvironment: "test",
+  environments: ["test", "prod"] as Array<string>,
```
Comment (slaesh) [inline apps/angular-example/src/app/abby.ts:8]: ?
```suggestion
  environments: ["test", "prod"],
```

File: apps/web/src/pages/api/config/[projectId]/index.ts [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+  const querySchemaResult = incomingQuerySchema.safeParse(req.query);
+
+  if (!querySchemaResult.success) {
+    res.status(400).json({ error: "Invalid Query Parameters" });
+    return;
+  }
+
+  const { projectId } = querySchemaResult.data;
+
+  const apiKey = req.headers["authorization"]?.split(" ")[1];
```
Comment (slaesh) [inline apps/web/src/pages/api/config/[projectId]/index.ts:42]: allowed headers says `Authorization`? is next somehow lower-casing them?

File: apps/web/src/pages/api/config/[projectId]/index.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+  transformClientFlagToDBType,
+  transformDBFlagTypeToclient,
+} from "lib/flags";
+import { AbbyConfigFile } from "@tryabby/core";
+
+const incomingQuerySchema = z.object({
+  projectId: z.string(),
+});
+
+export default async function handler(
```
Comment (slaesh) [inline apps/web/src/pages/api/config/[projectId]/index.ts:23]: this whole thingy is looking terrible. nested into the 1000th layer over hundreds of lines.

maybe we can use the beauty of modern programming languages: functions.

at least the individual "methods" GET/PUT..? and maybe even further..

File: apps/web/src/pages/profile.tsx [hunk 6]
```diff
@@ -34,59 +133,166 @@ const ProfilePage: NextPageWithLayout = () => {
...
+              </div>
+              <DashboardButton
+                type="submit"
+                className="ml-auto block px-5 py-1"
+              >
+                Save
+              </DashboardButton>
+            </form>
+          </section>
+          <section className="rounded-xl bg-gray-800 px-6 py-3">
```
Comment (slaesh) [inline apps/web/src/pages/profile.tsx:204]: maybe its time for dedicated components?

File: apps/web/src/pages/profile.tsx [hunk 7]
```diff
@@ -34,59 +133,166 @@ const ProfilePage: NextPageWithLayout = () => {
...
+        apiKey={newApiKeyInfo?.apiKey ?? ""}
+        name={newApiKeyInfo?.name ?? ""}
+        onClose={() => {
+          setIsCreateApiKeyModalOpen(false);
+          setNewApiKeyInfo(null);
+        }}
+      />
+      <RevokeApiKeyModal
+        isOpen={isRevokeApiKeyModalOpen}
+        apiKey={apiKeyToRevoke ?? undefined}
```
Comment (slaesh) [inline apps/web/src/pages/profile.tsx:289]: null to undefined? why do we use null at all?

File: apps/web/src/pages/projects/[projectId]/settings.tsx [hunk 8]
```diff
@@ -1,24 +1,27 @@
...
 import { getProjectPaidPlan, useAbbyStripe } from "lib/stripe";
+import { useSession } from "next-auth/react";
 import Link from "next/link";
 import { useRouter } from "next/router";
 import { NextPageWithLayout } from "pages/_app";
 import { FormEvent, useRef, useState } from "react";
 import { toast } from "react-hot-toast";
 import { BsX } from "react-icons/bs";
 import { getLimitByPlan } from "server/common/plans";
+import { generateRandomString } from "utils/apiKey";
```
Comment (slaesh) [inline apps/web/src/pages/projects/[projectId]/settings.tsx:23]: imports added, but no code changes? oO

File: apps/web/src/server/trpc/router/user.ts [hunk 9]
```diff
@@ -46,6 +46,17 @@ export const userRouter = router({
       }),
     };
   }),
+  getApiKeyData: protectedProcedure.query(async ({ ctx }) => {
+    return {
+      user: ctx.session.user.id,
```
Comment (slaesh) [inline apps/web/src/server/trpc/router/user.ts:51]: ?
```suggestion
      userId: ctx.session.user.id,
```

File: apps/web/src/utils/apiKey.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { createHmac, randomBytes } from "crypto";
+
+export function generateRandomString(length: number = 32): string {
```
Comment (slaesh) [inline apps/web/src/utils/apiKey.ts:3]:
```suggestion
export function generateRandomString(length = 32): string {
```

File: packages/cli/src/check.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+  const invalidFlags = Object.keys(remoteConfig.flags ?? {}).filter((key) => {
+    return (
+      localConfig.flags[key] == undefined &&
+      localConfig.flags[key] != (remoteConfig.flags ?? {})[key]
+    );
+  });
+
+  return {
+    isValid: invalidTests.length == 0 && invalidFlags.length == 0,
```
Comment (slaesh) [inline packages/cli/src/check.ts:36]:
```suggestion
    isValid: invalidTests.length === 0 && invalidFlags.length === 0,
```

File: packages/cli/src/http.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        body: JSON.stringify(localAbbyConfig),
+      });
+
+      const data = await response.json();
+      const status = response.status;
+
+      if (status == 200) {
+        console.log("pushed successfully");
+      } else if (status == 500) {
+        console.log("Pushed failed \n Please try again later \n 500: Internal server error");
```
Comment (slaesh) [inline packages/cli/src/http.ts:62]: if you want it more "readable, use multiline-strings or `[line1, line2, line3].join('\n')`
```suggestion
        console.log("Push failed\nPlease try again later\n500: Internal server error");
```

File: packages/cli/src/http.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+      } else if (status == 401) {
+        console.log("Pushed failed \n Please check your API key \n" + data);
```
Comment (slaesh) [inline packages/cli/src/http.ts:64]:
```suggestion
        console.log("Push failed\nPlease check your API key\n" + data);
```

File: packages/cli/src/http.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+      } else {
+        console.log("Pushed failed: \n" + status + ": " + data);
```
Comment (slaesh) [inline packages/cli/src/http.ts:66]: no one concatenates strings anymore :))

```suggestion
        console.log(`Push failed:\n${status}: data`);
```

File: packages/cli/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  .command("push")
+  .description("push local config to server")
+  .addOption(HostOption)
+  .addOption(ConfigOption)
+  .action(async (options: { config?: string; host?: string }) => {
+    try {
+      const token = await getToken();
+      await push({ apiKey: token, apiUrl: options.host, configPath: options.config });
+    } catch (e) {
+      console.log(chalk.red("Please login first"));
```
Comment (slaesh) [inline packages/cli/src/index.ts:60]: we tell the user in "any error case" to log in first? maybe this wasnt the reason? =)

File: packages/cli/src/pull.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+import { HttpService } from "./http";
+import { configRegex } from "./consts";
+import deepmerge from "deepmerge";
+import * as prettier from "prettier";
+
+export function updateConfigFile(updatedConfig: AbbyConfig, configFileString: string) {
+  const matchRegex = configRegex.exec(configFileString);
+
+  const matchedObject = matchRegex?.[1];
```
Comment (slaesh) [inline packages/cli/src/pull.ts:13]:
```suggestion
  const matchedObject = matchRegex?.at(1);
```

File: apps/angular-example/src/app/abby.ts (hunk 3 above)
Comment (cstrnt) [reply apps/angular-example/src/app/abby.ts:8]: I need to change this file around. The variable is cast as const and therefore the env will be infered to `ReadOnlyArray<string>` which isn't what createAbby takes in as an argument. Good catch tho!

File: apps/web/src/pages/api/config/[projectId]/index.ts (hunk 4 above)
Comment (cstrnt) [reply apps/web/src/pages/api/config/[projectId]/index.ts:42]: yup,  that's whats happening :(

File: apps/web/src/pages/projects/[projectId]/settings.tsx (hunk 8 above)
Comment (cstrnt) [reply apps/web/src/pages/projects/[projectId]/settings.tsx:23]: What do you mean?

File: apps/web/src/pages/api/config/[projectId]/index.ts (hunk 5 above)
Comment (cstrnt) [reply apps/web/src/pages/api/config/[projectId]/index.ts:23]: I agree, makes sense to break this up. Newer next versions actually allowing exporting individual functions such a `GET` or `PUT` which makes it much cleaner

File: apps/web/src/pages/profile.tsx (hunk 6 above)
Comment (cstrnt) [reply apps/web/src/pages/profile.tsx:204]: yup

File: apps/web/src/pages/profile.tsx (hunk 7 above)
Comment (cstrnt) [reply apps/web/src/pages/profile.tsx:289]: Good question about a philosophical question of javascript. I like having `null` to verbosely specify the explicit abstinence of a value. `undefined` is implicit version. I get why people like it the other way tho

File: apps/web/src/pages/projects/[projectId]/settings.tsx (hunk 8 above)
Comment (cstrnt) [reply apps/web/src/pages/projects/[projectId]/settings.tsx:23]: Oh! Import added and not used?

File: packages/cli/src/http.ts (hunk 12 above)
Comment (cstrnt) [reply packages/cli/src/http.ts:60]: good catch

File: apps/web/src/pages/profile.tsx (hunk 7 above)
Comment (slaesh) [reply apps/web/src/pages/profile.tsx:289]: even JS uses undefined > null. array.find => undefined.

Comment (slaesh) [reply apps/web/src/pages/profile.tsx:289]: so.. if you want to use null, do it everywhere! consistency is the key.

File: apps/docs/pages/integrations/angular.mdx [hunk 17]
```diff
@@ -34,28 +34,38 @@ To get started make sure to install the packages using your favorite package man
...
+export default defineConfig({
+  projectId: "<YOUR_PROJECT_ID>",
+  currentEnvironment: environment.production ? "production" : "development",
+  environments: ["production", "development"],
+  tests: {
+    test: { variants: ["A", "B"] },
+    footer: { variants: ["dark", "orange", "green"] },
+    // ... your tests
+  },
+  flags: { darkMode: "Boolean", newFeature: "Boolean" },
```
Comment (slaesh) [inline apps/docs/pages/integrations/angular.mdx:54]: where and how can I define my default values? :thinking:

File: apps/angular-example/src/app/app.module.ts [hunk 18]
```diff
@@ -1,16 +1,16 @@
-import { APP_INITIALIZER, Injectable, NgModule } from '@angular/core';
-import { BrowserModule } from '@angular/platform-browser';
+import { APP_INITIALIZER, Injectable, NgModule } from "@angular/core";
+import { BrowserModule } from "@angular/platform-browser";
```
Comment (slaesh) [inline apps/angular-example/src/app/app.module.ts:3]:
```suggestion
```

File: apps/web/src/components/CreateAPIKeyModal.tsx [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    onSuccess() {
+      toast.success("API Key created");
+    },
+  });
+
+  return (
+    <Modal
+      isOpen={isOpen}
+      onClose={onClose}
+      title="Create new Api Key"
```
Comment (slaesh) [inline apps/web/src/components/CreateAPIKeyModal.tsx:30]:
```suggestion
      title="Create new API Key"
```

File: apps/web/src/components/FeatureFlag.tsx [hunk 20]
```diff
@@ -200,7 +200,7 @@ export function FeatureFlag({
 
   const currentFlagValue = flag.values.find((f) => f.id === flagValueId)?.value;
 
-  if (!currentFlagValue) {
+  if (currentFlagValue === undefined) {
```
Comment (slaesh) [inline apps/web/src/components/FeatureFlag.tsx:203]: couldn't `undefined` be a valid value? why not?

File: apps/web/src/components/FeatureFlag.tsx [hunk 21]
```diff
@@ -213,7 +213,10 @@ export function FeatureFlag({
             title={currentFlagValue}
             className="max-w-[60px] overflow-hidden text-ellipsis whitespace-nowrap rounded-md bg-gray-600 p-1"
           >
-            {currentFlagValue}
+            {typeof currentFlagValue === "string" &&
```
Comment (slaesh) [inline apps/web/src/components/FeatureFlag.tsx:216]: i am no react expert.. but adding "so much" code here seems odd? cant we just have a variable like "trimmedStringValue" or so? to easily acces it here? or even better, "isValueEmptyString"

File: packages/angular/src/lib/abby.service.ts [hunk 22]
```diff
@@ -114,13 +123,13 @@ export class AbbyService<
...
       map((variant) => {
-        if(lookupObject === undefined) {
+        if (lookupObject === undefined) {
           return variant;
         }
 
         return lookupObject[variant];
       }),
-      shareReplay(1),
+      shareReplay(1)
```
Comment (slaesh) [inline packages/angular/src/lib/abby.service.ts:132]: guess the project needs some sort of setup for this formatting ping-pong

File: packages/cli/src/check.ts [hunk 23]
```diff
@@ -0,0 +1,40 @@
+import { HttpService } from "./http";
+import { loadLocalConfig } from "./util";
+
+export async function check({
```
Comment (slaesh) [inline packages/cli/src/check.ts:4]: check? check what :) the weather? ^^ there is no limit in naming functions..
even after looking into it its not clear what is going on here.. what is being checked. what is invalid in the return? the local one, the remote one?

File: packages/cli/src/pull.ts [hunk 24]
```diff
@@ -0,0 +1,72 @@
+import * as fs from "fs/promises";
+import { AbbyConfig, PullAbbyConfigResponse } from "@tryabby/core";
+import { loadLocalConfig } from "./util";
+
```
Comment (slaesh) [inline packages/cli/src/pull.ts:4]:
```suggestion
```

File: apps/web/src/components/FeatureFlag.tsx (hunk 20 above)
Comment (cstrnt) [reply apps/web/src/components/FeatureFlag.tsx:203]: We don't allow undefined as a value right now.

It's either an empty string, `false`, the integer `0` or an empty json object `{}`

File: apps/docs/pages/integrations/angular.mdx (hunk 17 above)
Comment (cstrnt) [reply apps/docs/pages/integrations/angular.mdx:54]: Good catch, it's missing in the angular docs

File: apps/web/src/components/FeatureFlag.tsx (hunk 21 above)
Comment (cstrnt) [reply apps/web/src/components/FeatureFlag.tsx:216]: I don't really think it's odd to have it here. The check is only done once and therefore I rather not have a variable which is in-fact only used once. I think this comes down to personal preference

File: packages/angular/src/lib/abby.service.ts (hunk 22 above)
Comment (cstrnt) [reply packages/angular/src/lib/abby.service.ts:132]: yes, 100%

File: packages/cli/src/check.ts (hunk 23 above)
Comment (cstrnt) [reply packages/cli/src/check.ts:4]: Good catch!

File: apps/web/src/components/FeatureFlag.tsx (hunk 20 above)
Comment (slaesh) [reply apps/web/src/components/FeatureFlag.tsx:203]: but `null` is okay?
are we checking here the result of the "find" function or could the actual `value` be undefined/null? if so, maybe split the check into two?

Comment (cstrnt) [reply apps/web/src/components/FeatureFlag.tsx:203]: you're right again, thanks :)

**Review by slaesh — APPROVED** [review]

**Pull request merged**